File: services/pipeline/internal/contract/fixtures_test.go
package contract

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"slices"
	"strings"
	"testing"
	"time"
)

const fixtureDirectory = "../../../../packages/contract/fixtures"

func fixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(fixtureDirectory, name))
	if err != nil {
		t.Fatal(err)
	}
	return data
}

func TestFixtureRoundTrips(t *testing.T) {
	fixtures := map[string]func() any{
		"events.batch.json":           func() any { return new(EventBatch) },
		"healthz.response.json":       func() any { return new(HealthResponse) },
		"ingest.hello.json":           func() any { return new(IngestHello) },
		"ingest.ready.json":           func() any { return new(IngestReady) },
		"ingest.stats.json":           func() any { return new(IngestStats) },
		"session-start.request.json":  func() any { return new(SessionStartRequest) },
		"session-start.response.json": func() any { return new(SessionStartResponse) },
		"sessions.response.json":      func() any { return new(SessionsResponse) },
	}
	entries, err := os.ReadDir(fixtureDirectory)
	if err != nil {
		t.Fatal(err)
	}
	var actual []string
	for _, entry := range entries {
		actual = append(actual, entry.Name())
	}
	var expected []string
	for name := range fixtures {
		expected = append(expected, name)
	}
	expected = append(expected, "ingest-token.vector.json")
	slices.Sort(actual)
	slices.Sort(expected)
	if !slices.Equal(actual, expected) {
		t.Fatalf("fixtures without Go types: got %v, want %v", actual, expected)
	}

	for name, newMessage := range fixtures {
		t.Run(name, func(t *testing.T) {
			original := fixture(t, name)
			message := newMessage()
			if err := json.Unmarshal(original, message); err != nil {
				t.Fatal(err)
			}
			encoded, err := json.Marshal(message)
			if err != nil {
				t.Fatal(err)
			}
			var want, got any
			if err := json.Unmarshal(original, &want); err != nil {
				t.Fatal(err)
			}
			if err := json.Unmarshal(encoded, &got); err != nil {
				t.Fatal(err)
			}
			if !reflect.DeepEqual(got, want) {
				t.Fatalf("round trip differs: got %s, want %s", encoded, original)
			}
		})
	}
}

func TestEventVariants(t *testing.T) {
	var batch EventBatch
	if err := json.Unmarshal(fixture(t, "events.batch.json"), &batch); err != nil {
		t.Fatal(err)
	}
	if batch.ContractVersion != Version || len(batch.Events) != 4 || batch.Events[0].Segment == nil || batch.Events[1].Segment == nil || batch.Events[2].Status == nil || batch.Events[3].Log == nil {
		t.Fatalf("unexpected event variants: %+v", batch)
	}
	for _, input := range []string{
		`{"type":"unknown"}`,
		`{"type":"segment","unexpected":true}`,
	} {
		var event Event
		if err := json.Unmarshal([]byte(input), &event); err == nil {
			t.Fatalf("accepted invalid event: %s", input)
		}
	}
}

func TestIngestTokenVector(t *testing.T) {
	var vector struct {
		Secret        string `json:"secret"`
		Payload       string `json:"payload"`
		ExpectedToken string `json:"expectedToken"`
	}
	if err := json.Unmarshal(fixture(t, "ingest-token.vector.json"), &vector); err != nil {
		t.Fatal(err)
	}
	var claims IngestTokenClaims
	payload, err := base64.RawURLEncoding.DecodeString(vector.Payload)
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(payload, &claims); err != nil {
		t.Fatal(err)
	}
	key, err := base64.StdEncoding.DecodeString(vector.Secret)
	if err != nil {
		t.Fatal(err)
	}
	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(vector.Payload))
	if got := vector.Payload + "." + base64.RawURLEncoding.EncodeToString(mac.Sum(nil)); got != vector.ExpectedToken {
		t.Fatalf("vector mismatch: got %s", got)
	}

	validAt := time.Unix(claims.Exp-1, 0)
	if !VerifyIngestToken(vector.ExpectedToken, claims.SessionID, vector.Secret, validAt) {
		t.Fatal("rejected shared ingest token vector")
	}
	if VerifyIngestToken(vector.ExpectedToken, claims.SessionID, vector.Secret, time.Unix(claims.Exp, 0)) {
		t.Fatal("accepted expired token")
	}
	if VerifyIngestToken(vector.ExpectedToken, "5c3b3b4e-1c1e-4a2e-9f0d-9a3f5b1e2d77", vector.Secret, validAt) {
		t.Fatal("accepted token for a different session")
	}
	parts := strings.Split(vector.ExpectedToken, ".")
	for _, token := range []string{
		"malformed",
		parts[0] + ".A" + parts[1][1:],
		parts[0] + "." + parts[1][:len(parts[1])-1] + "5", // same bytes, non-canonical pad bits
	} {
		if VerifyIngestToken(token, claims.SessionID, vector.Secret, validAt) {
			t.Fatalf("accepted malformed token: %s", token)
		}
	}
	if VerifyIngestToken(vector.ExpectedToken, claims.SessionID, "invalid", validAt) {
		t.Fatal("accepted invalid shared secret")
	}
}

File: apps/web/src/lib/contract/fixtures.test.ts
import { readdirSync, readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import type { ZodTypeAny } from "zod";

import {
	contractVersion,
	eventBatchSchema,
	healthResponseSchema,
	ingestHelloSchema,
	ingestReadySchema,
	ingestStatsSchema,
	sessionStartRequestSchema,
	sessionStartResponseSchema,
	sessionsResponseSchema,
} from "./index";
import { mintIngestToken, verifyIngestToken } from "./token.server";

const fixturesDirectory = new URL(
	"../../../../../packages/contract/fixtures/",
	import.meta.url,
);

function fixture(name: string): unknown {
	return JSON.parse(readFileSync(new URL(name, fixturesDirectory), "utf8"));
}

const fixtureSchemas: Record<string, ZodTypeAny> = {
	"events.batch.json": eventBatchSchema,
	"healthz.response.json": healthResponseSchema,
	"ingest.hello.json": ingestHelloSchema,
	"ingest.ready.json": ingestReadySchema,
	"ingest.stats.json": ingestStatsSchema,
	"session-start.request.json": sessionStartRequestSchema,
	"session-start.response.json": sessionStartResponseSchema,
	"sessions.response.json": sessionsResponseSchema,
};

describe("contract fixtures", () => {
	it("has a schema for every message fixture", () => {
		expect(readdirSync(fixturesDirectory).sort()).toEqual(
			[...Object.keys(fixtureSchemas), "ingest-token.vector.json"].sort(),
		);
	});

	it.each(Object.entries(fixtureSchemas))("round-trips %s", (name, schema) => {
		const original = fixture(name);
		const parsed = schema.parse(original);
		expect(JSON.parse(JSON.stringify(parsed))).toEqual(original);
	});

	it("rejects an incompatible version and malformed event", () => {
		expect(contractVersion).toBe(2);
		const batch = fixture("events.batch.json") as {
			contractVersion: number;
			events: Array<Record<string, unknown>>;
		};
		expect(
			eventBatchSchema.safeParse({ ...batch, contractVersion: 3 }).success,
		).toBe(false);
		expect(
			eventBatchSchema.safeParse({
				...batch,
				events: [{ ...batch.events[0], startMs: -1 }],
			}).success,
		).toBe(false);
	});

	it("keeps optional speaker labels and rejects an empty one", () => {
		const batch = eventBatchSchema.parse(fixture("events.batch.json"));
		const segments = batch.events.filter((event) => event.type === "segment");
		expect(segments[0]?.speaker).toBe("S1");
		expect(segments[1]?.speaker).toBe("S1");
		expect(
			eventBatchSchema.safeParse({
				...batch,
				events: [{ ...batch.events[0], speaker: "" }],
			}).success,
		).toBe(false);
		expect(
			eventBatchSchema.safeParse({
				...batch,
				events: [{ ...batch.events[0], speaker: undefined }],
			}).success,
		).toBe(true);
	});
});

describe("ingest tokens", () => {
	const vector = fixture("ingest-token.vector.json") as {
		secret: string;
		payload: string;
		expectedToken: string;
	};
	const claims = JSON.parse(
		Buffer.from(vector.payload, "base64url").toString("utf8"),
	) as {
		sessionId: string;
		exp: number;
	};
	const mintedAt = claims.exp - 600;

	it("reproduces the shared vector byte for byte", () => {
		expect(mintIngestToken(claims.sessionId, vector.secret, mintedAt)).toBe(
			vector.expectedToken,
		);
		expect(
			verifyIngestToken(
				vector.expectedToken,
				claims.sessionId,
				vector.secret,
				mintedAt,
			),
		).toBe(true);
	});

	it("rejects expiry, a different session, and a changed signature", () => {
		expect(
			verifyIngestToken(
				vector.expectedToken,
				claims.sessionId,
				vector.secret,
				claims.exp,
			),
		).toBe(false);
		expect(
			verifyIngestToken(
				vector.expectedToken,
				"5c3b3b4e-1c1e-4a2e-9f0d-9a3f5b1e2d77",
				vector.secret,
				mintedAt,
			),
		).toBe(false);
		const [payload, signature] = vector.expectedToken.split(".");
		expect(
			verifyIngestToken(
				`${payload}.A${signature?.slice(1)}`,
				claims.sessionId,
				vector.secret,
				mintedAt,
			),
		).toBe(false);
		const alphabet =
			"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
		const finalCharacter = signature?.at(-1) ?? "";
		const index = alphabet.indexOf(finalCharacter);
		const alternate = alphabet[(index & ~3) | ((index + 1) & 3)];
		expect(
			verifyIngestToken(
				`${payload}.${signature?.slice(0, -1)}${alternate}`,
				claims.sessionId,
				vector.secret,
				mintedAt,
			),
		).toBe(false);
		expect(
			verifyIngestToken("malformed", claims.sessionId, vector.secret, mintedAt),
		).toBe(false);
	});
});
